### PR TITLE
Add fwupd_guid_from_string() to drop dep on uuid

### PR DESCRIPTION
**Pull request by hughsie**
This also allows us to write mixed-endian structures and adds tests. As part of
this commit we've also changed the API of something that's not yet been in any
tarball release, so no pitchforks please.

Type of pull request:
- [ ] New plugin (Please include [new plugin checklist](https://github.com/hughsie/fwupd/wiki/New-plugin-checklist))
- [x] Code fix
- [x] Feature
- [ ] Documentation

### CODE REVIEW

**Comment by hughsie** [comment]
Also, I don't get the CentOS failure, advice welcome:

    ../libfwupd/fwupd-common.h:52:9: note: expected ‘const guint8 (*)[16]’ but argument is of type ‘guint8 (*)[16]’

I thought it was always valid to pass `TYPE` to a function expecting `CONST TYPE`?

File: libfwupd/fwupd-common.c [hunk 1]
```diff
@@ -537,39 +536,187 @@ typedef struct __attribute__((packed)) {
...
+			     G_IO_ERROR, G_IO_ERROR_INVALID_DATA,
+			     "Number “%s” is out of bounds [%s, %s]",
+			     str, min_str, max_str);
+		return FALSE;
+	}
+	if (out_num != NULL)
+		*out_num = number;
+	return TRUE;
+}
+#endif
```
Comment (superm1) [inline libfwupd/fwupd-common.c]: maybe add  /* GLIB_CHECK_VERSION(2,54,0) */

File: libfwupd/fwupd-common.c [hunk 2]
```diff
@@ -537,39 +536,187 @@ typedef struct __attribute__((packed)) {
...
+{
+	const gchar *end_ptr = NULL;
+	gint saved_errno = 0;
+	guint64 number;
+
+	g_return_val_if_fail (str != NULL, FALSE);
+	g_return_val_if_fail (base >= 2 && base <= 36, FALSE);
+	g_return_val_if_fail (min <= max, FALSE);
+	g_return_val_if_fail (error == NULL || *error == NULL, FALSE);
+
```
Comment (superm1) [inline libfwupd/fwupd-common.c]: I'd think you also want a check for `out_num` here no?

File: libfwupd/fwupd-common.c [hunk 3]
```diff
@@ -633,20 +778,17 @@ fwupd_guid_from_data (const guint8 *data, gsize datasz, FwupdGuidFlags flags)
 gboolean
 fwupd_guid_is_valid (const gchar *guid)
 {
-	gint rc;
-	uuid_t uu;
 	if (guid == NULL)
 		return FALSE;
-	if (strlen (guid) != 36)
+	if (!fwupd_guid_from_string (guid, NULL, FWUPD_GUID_FLAG_NONE, NULL))
```
Comment (superm1) [inline libfwupd/fwupd-common.c]: Don't you want to check what the error was for not being valid?  Or just don't care?

File: libfwupd/fwupd-common.c (hunk 1 above)
Comment (hughsie) [reply libfwupd/fwupd-common.c]: can do

File: libfwupd/fwupd-common.c (hunk 2 above)
Comment (hughsie) [reply libfwupd/fwupd-common.c]: I just copied it from GLib :) I'm tempted to just fix it upstream for things like this. In an ideal world I'd bump the GLib dep, but that might make me unpopular...

Comment (superm1) [reply libfwupd/fwupd-common.c]: Yeah, I think holding off as long as possible on bumping that particular dep is ideal.

File: libfwupd/fwupd-common.c [hunk 4]
```diff
@@ -537,39 +536,187 @@ typedef struct __attribute__((packed)) {
...
+			fwupd_guid_t *guid,
+			FwupdGuidFlags flags,
+			GError **error)
+{
+	fwupd_guid_native_t gu = { 0x0 };
+	gboolean mixed_endian = flags & FWUPD_GUID_FLAG_MIXED_ENDIAN;
+	guint64 tmp;
+	g_auto(GStrv) split = NULL;
+
+	g_return_val_if_fail (guidstr != NULL, FALSE);
```
[the remark this answers is not in the document]
Comment (hughsie) [reply libfwupd/fwupd-common.c]: Na, `guid` can be NULL deliberately to just check the format. We do:

    if (guid != NULL)
        memcpy (guid, &gu, sizeof(gu));

File: libfwupd/fwupd-common.c (hunk 3 above)
Comment (hughsie) [reply libfwupd/fwupd-common.c]: I'd say just don't care. I suppose we could add a GError if you think it's a good idea.

Comment (superm1) [reply libfwupd/fwupd-common.c]: Agree Since you have a deliberate NULL case just for checking format, seems like needless work and debugging output.

**Comment by hughsie** [comment]
> I do wonder if it's worth dropping usage of `efi_guid_t` in the UEFI plugin now too.

I pondered that. I might do that as a follow-up -- it's not like we pass fwupd_guid_t to `fu_device_add_guid()` and we get `efi_guid_t` for "free" from libefivar.

**Pull request merged**